Fix concat not respecting order of OrderedDict

alexander-ponomaroff opened the pull request:
- [x] closes #21510
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

alexander-ponomaroff committed:
Fix concat not respecting order of OrderedDict

alexander-ponomaroff committed:
Merge conflict fix

pep8speaks commented:
Hello @alexander-ponomaroff! Thanks for submitting the PR.





 - In the file [`pandas/tests/frame/test_constructors.py`](https://github.com/pandas-dev/pandas/blob/073ade88e0234e0043c9e37e81f32e37e897dec1/pandas/tests/frame/test_constructors.py), following are the PEP8 issues :

> [Line 15:80](https://github.com/pandas-dev/pandas/blob/073ade88e0234e0043c9e37e81f32e37e897dec1/pandas/tests/frame/test_constructors.py#L15): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

alexander-ponomaroff committed:
Fix pep8 issues

alexander-ponomaroff commented:
@jreback @toobaz I've reviewed your comments in Pull Request #21512 that went stale, and created a fix based on what I observed. As @toobaz said, the ordering of dicts in 3.6 is an implementation detail of cpython, and users should not rely on it in 3.6. Based on this, I changed dict_keys_to_ordered_list to use PY37 instead of PY36 and changed the tests that were failing to use PY37 as well.

Please let me know if this is the wrong way of going about this. If my changes are acceptable, I will change the comments in the tests and dict_keys_to_ordered_list to use 3.7 instead of 3.6.

alexander-ponomaroff committed:
Fix import order

WillAyd commented:
> As @toobaz said, the ordering of dicts in 3.6 is an implementation detail of cpython, and users should not rely on it in 3.6

Might be out of the loop but wasn't it declared that at first but since then guaranteed as a feature? If so I don't think we need to make distinction between 3.6 and 3.7 here

jreback commented on pandas/core/common.py:
```diff
@@ -226,7 +226,7 @@ def try_sort(iterable):
     # when pandas drops support for Python < 3.6, this function
     # can be replaced by a simple list(mapping.keys())
-    if PY36 or isinstance(mapping, OrderedDict):
+    if PY37 or isinstance(mapping, OrderedDict):
```
pls change these back to PY36 which is fine.

jreback reviewed (changes requested):
pls add a whatsnew note to 0.25 in reshaping.

it may be an implementation detail in PY36 but we are consistently testing for this.

alexander-ponomaroff commented:
@jreback Thank you for the feedback, I will make changes tomorrow.

alexander-ponomaroff committed:
Checks reversed to PY36

alexander-ponomaroff committed:
Merge conflict

alexander-ponomaroff committed:
Forgot to remove PY37 from imports

alexander-ponomaroff commented:
@WillAyd I get an error in one of the tests that you created. Would you have any suggestions for me? test_groupby_agg_ohlc_non_first() inside pandas/tests/groupby/test_groupby.py.

WillAyd commented:
Looks like its the ordering of the columns in the resulting data frame. I think this is actually preferable because it respects the order supplied by the user, so I think OK to update test accordingly

alexander-ponomaroff committed:
Change order of expected result in test_groupby_agg_ohlc_non_first() and remove an unwanted import

alexander-ponomaroff commented:
@WillAyd Thank you for such a fast response. Just to confirm, how do you feel about the change I made?

alexander-ponomaroff committed:
Fixed accidental empty line removal

alexander-ponomaroff committed:
Fix test_groupby_agg_ohlc_non_first()

WillAyd commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1695,6 +1695,10 @@ def test_groupby_agg_ohlc_non_first():
         ('foo', 'sum', 'foo'))), index=pd.date_range(
             '2018-01-01', periods=2, freq='D'))
 
+    if compat.PY36:
```
Hmm I wasn't thinking about this on original reply. Others may disagree but I don't think we want to return a different order depending on Python version. Do you know the code responsible for this behavior?

alexander-ponomaroff replied:
@WillAyd What do you mean by "Do you know the code responsible for this behavior?". Do I know what is responsible for different order of columns in the DataFrame in the particular case of this test?

TomAugspurger replied:
@WillAyd was this question resolved?

I think that @WillAyd's question was why does the order of `result` depend on the python version?

WillAyd replied:
Yea I think this is still outstanding

alexander-ponomaroff replied:
@WillAyd Your latest suggestion included this:

> When doing .agg(['ohlc', 'sum']), it passes the test in all the versions
> 
> My point was to just change the test to this instead of doing the compat. Should work in all versions then, no?

To which I responded:

> I did that in one of my commits, however I realized that the whole point of the test was to test ohlc when it's not passed in first into the agg function. There used to be an error when ohlc wasn't passed in first, but you fixed that. So the test was created to ensure that everything will still work even if ohlc isn't first, so by changing it to agg(['ohlc', 'sum']) will kind of defeat the purpose of the test, because ohlc is first. If changing it to agg(['ohlc', 'sum']) is good with you, I will do it, but I just wanted to bring this up before I do it.

Should I still change the test to .agg(['ohlc', 'sum']), which will pass in all versions? Or do you have another suggestion in mind?

WillAyd replied:
No - thanks for highlighting that about the initial purpose of the test. I guess I'm unclear as to why this change would introduce a difference in the ordering of the result across versions - shouldn't it be the same?

alexander-ponomaroff replied:
@WillAyd To my understanding it should be different because previous to Python 3.6 it would sort what the user passed into the `agg` function. But since I introduced the change that would change the order of the `objs` in the `concat` function based on the Python version and `concat` is used within the .agg(['sum', 'ohlc']) call, the order of the output will be different based on the Python version (sorted in <3.6 and maintain the order that user entered in >=3.6).

WillAyd replied:
So I believe this is the offending line in the groupby code:

https://github.com/pandas-dev/pandas/blob/07625af622c650f2ea124e3ddc5a80df37e6a2aa/pandas/core/groupby/generic.py#L788

The dict which gets passed into that concat call is built here:

https://github.com/pandas-dev/pandas/blob/07625af622c650f2ea124e3ddc5a80df37e6a2aa/pandas/core/groupby/generic.py#L825

I'm not sure of best approach to handle. One one hand maybe we convert all of the groupby code to OrderedDict to prevent this, but at the same time I'm not sure why this would appear as part of this PR if the focus is only on OrderedDict, which isn't in use here.

Any idea where that impact may be stemming from within this?

alexander-ponomaroff replied:
@WillAyd This PR also deals with regular dicts in Python 3.6+ because they should exhibit the same behaviour as OrderedDict in Python 3.6+. So this PR ensures that order of user input in the concat function is maintained for OrderedDicts in all versions and regular dicts in Python 3.6+. That's why regular dicts are also affected based on the version. With this in mind, do you think that we should convert the dict that you specified in the above comment to an OrderedDict for the result to be the same in all versions?

jreback replied:
@alexander-ponomaroff if you change the ``results = OrderedDict()`` in groupby/generic.py as above. you won't need to change this test code.

TomAugspurger replied:
@alexander-ponomaroff can you update?

I don't think that this PR should require changing the test. Rather, we'll update groupby to ensure that the columns are consistent with the previous results for all versions of Python.

WillAyd commented:
Right - I’m asking if there’s a way to keep the behavior here consistent in the code base rather than setting different expectations in the test

> On Feb 12, 2019, at 2:40 PM, Alexander Ponomaroff <notifications@github.com> wrote:
> 
> @alexander-ponomaroff commented on this pull request.
> 
> In pandas/tests/groupby/test_groupby.py <https://github.com/pandas-dev/pandas/pull/25224#discussion_r256177741>:
> 
> > @@ -1695,6 +1695,10 @@ def test_groupby_agg_ohlc_non_first():
>          ('foo', 'sum', 'foo'))), index=pd.date_range(
>              '2018-01-01', periods=2, freq='D'))
>  
> +    if compat.PY36:
> @WillAyd <https://github.com/WillAyd> What do you mean by "Do you know the code responsible for this behavior?". Do I know what is responsible for different order of columns in the DataFrame in the particular case of this test?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/pandas-dev/pandas/pull/25224#discussion_r256177741>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAlOUQ6ufDxF8qxcIyIHxcBpfSfu9SR_ks5vM0LEgaJpZM4attTh>.
>

alexander-ponomaroff commented:
@WillAyd I will have to investigate this and get back to you. I'm not entirely too sure how my change affected test_groupby_agg_ohlc_non_first() because my changes deal with dicts and OrderedDicts in the concat function. So, concat must be used within one of the functions that's used inside the test, but the test doesn't seem to be dealing with any dicts. Please let me know if anything I'm saying is wrong, I recently started solving actual code bugs in Pandas and am still learning. I started my contributions with easy issues and now trying to move onto harder ones :)

WillAyd commented:
Yea no worries - your contributions are greatly appreciated.

If it helps here is where the ohlc function is defined. If you trace the underlying function there back you'll see a concat call, hence why this has some impacts

https://github.com/pandas-dev/pandas/blob/4d44a2aff08c3075d9472e361bd7448a0472494e/pandas/core/groupby/groupby.py#L1311

alexander-ponomaroff commented:
@WillAyd Thank you. I will be looking into it and will update when I find something.

alexander-ponomaroff committed:
Testing if this will pass tests in <3.6

alexander-ponomaroff committed:
Merge

alexander-ponomaroff commented:
@WillAyd I've been trying to figure out how to solve the problem with test_groupby_agg_ohlc_non_first() but I'm stuck. Python 3.6 and after, .agg(['sum', 'ohlc']) maintains the order "sum -> ohlc", .agg(['ohlc', 'sum']) maintains the order "ohlc -> sum". When doing .agg(['ohlc', 'sum']), it passes the test in all the versions. This is due to my recent concat change and I traced the call and get why it happens, but I don't know how to solve the issue in this test, without affecting other tests negatively.

WillAyd commented:
If that’s the case I think it’s preferable to change the test rather than add the compat condition. Same thing at the end of the day just less clutter 

Sent from my iPhone

> On Feb 13, 2019, at 11:29 PM, Alexander Ponomaroff <notifications@github.com> wrote:
> 
> @WillAyd I've been trying to figure out how to solve the problem with test_groupby_agg_ohlc_non_first() but I'm stuck. Python 3.6 and after, .agg(['sum', 'ohlc']) maintains the order "sum -> ohlc", .agg(['ohlc', 'sum']) maintains the order "ohlc -> sum". When doing .agg(['ohlc', 'sum']), it passes the test in all the versions. This is due to my recent concat change and I traced the call and get why it happens, but I don't know how to solve the issue in this test, without affecting other tests negatively.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

alexander-ponomaroff commented:
@WillAyd But the result of .agg(['sum', 'ohlc']) will be different based on the version. So without checking for the version with compat and without changing the order in agg, how can the test be rewritten?

alexander-ponomaroff committed:
Revert back to compat check for now

alexander-ponomaroff committed:
Merge master

WillAyd commented:
> When doing .agg(['ohlc', 'sum']), it passes the test in all the versions

My point was to just change the test to this instead of doing the compat. Should work in all versions then, no?

alexander-ponomaroff commented:
@WillAyd I did that in one of my commits, however I realized that the whole point of the test was to test ohlc when it's not passed in first into the agg function. There used to be an error when ohlc wasn't passed in first, but you fixed that. So the test was created to ensure that everything will still work even if ohlc isn't first, so by changing it to agg(['ohlc', 'sum']) will kind of defeat the purpose of the test, because ohlc is first. If changing it to agg(['ohlc', 'sum']) is good with you, I will do it, but I just wanted to bring this up before I do it.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -183,6 +183,7 @@ Reshaping
 - Bug in :func:`pandas.merge` adds a string of ``None`` if ``None`` is assigned in suffixes instead of remain the column name as-is (:issue:`24782`).
 - Bug in :func:`merge` when merging by index name would sometimes result in an incorrectly numbered index (:issue:`24212`)
 - :func:`to_records` now accepts dtypes to its `column_dtypes` parameter (:issue:`24895`)
+- Bug in :func:`concat` order of OrderedDict is not respected (:issue:`21510`)
```
Bug in :func:`concat` where the order of a ``OrderedDict`` is not respected (use double backticks on OrderedDict)

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1695,6 +1695,10 @@ def test_groupby_agg_ohlc_non_first():
 
+    if compat.PY36:
+        sorted_keys = sorted(expected, key=lambda x: x[1], reverse=True)
+        expected = expected.reindex(columns=sorted_keys)
```
instead of thisI would just sort the columns in above

jreback reviewed (changes requested):
small comments, ping on green.

alexander-ponomaroff committed:
Requested changes

alexander-ponomaroff committed:
Merge master

alexander-ponomaroff commented:
@jreback Why am I failing these tests now? Does it have anything to do with changes happened in the last 2 days?

jreback commented:
make sure to merge master
these were just fixed up today

alexander-ponomaroff commented:
@jreback I did merge master, right before I pushed my changes. It says that everything is up to date.

jreback commented:
these look related to your changes
run the tests locally and see

alexander-ponomaroff committed:
Merge

alexander-ponomaroff committed:
Merge

alexander-ponomaroff commented:
@jreback All test are green now. 

Looks like the previous errors were resolved by themselves. When I ran the test locally, I did not get them.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -185,6 +185,7 @@ Reshaping
 - Bug in :func:`pandas.merge` adds a string of ``None`` if ``None`` is assigned in suffixes instead of remain the column name as-is (:issue:`24782`).
 - Bug in :func:`merge` when merging by index name would sometimes result in an incorrectly numbered index (:issue:`24212`)
 - :func:`to_records` now accepts dtypes to its `column_dtypes` parameter (:issue:`24895`)
+- Bug in :func:`concat` order of ``OrderedDict`` is not respected (:issue:`21510`)
```
where order of ``OrderedDict``

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -1163,7 +1163,7 @@ def test_concat_dict(self):
                   'qux': DataFrame(np.random.randn(4, 3))}
 
-        sorted_keys = sorted(frames)
+        sorted_keys = pd.core.common.dict_keys_to_ordered_list(frames)
```
import this at the top instead

jreback reviewed (changes requested):
pls merge master as well

alexander-ponomaroff committed:
Remaining small requested fixes

alexander-ponomaroff committed:
Merge

alexander-ponomaroff committed:
Fix unwanted import error

alexander-ponomaroff commented:
@jreback Fixed requested changes, all tests are green.

TomAugspurger commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -241,6 +241,7 @@ Reshaping
 - Bug in :func:`pandas.merge` adds a string of ``None`` if ``None`` is assigned in suffixes instead of remain the column name as-is (:issue:`24782`).
 - Bug in :func:`merge` when merging by index name would sometimes result in an incorrectly numbered index (:issue:`24212`)
 - :func:`to_records` now accepts dtypes to its `column_dtypes` parameter (:issue:`24895`)
+- Bug in :func:`concat` where order of ``OrderedDict`` is not respected (:issue:`21510`)
```
Based just on the release note, I have two questions:

1. What argument is being passed as an OrderedDict here?
2. Does this change apply to regular dicts in Python 3.6+?

alexander-ponomaroff replied:
1. Do you mean which argument in the concat function is being passed as an OrderedDict?
2. Yes, because I make use of a function that also checks for dicts that are in Python 3.6+ and they follow the same order as the OrderedDict.

TomAugspurger replied:
1. Yes. Presumably `objs`, but a reader may potentially think `keys`.
2. It'd be good to mention that in a parenthetical.

alexander-ponomaroff replied:
Bug in :func:`concat` where order of ``OrderedDict`` (and ``dict`` in Python 3.6+) is not respected, when passed in as  ``objs`` argument (:issue:`21510`)

Would you suggest changing the release note to this?

jreback replied:
yes

alexander-ponomaroff committed:
Use OrderedDict in groupby/generic.py

alexander-ponomaroff committed:
Merge

TomAugspurger commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1690,14 +1690,11 @@ def test_groupby_agg_ohlc_non_first():
-        ('foo', 'ohlc', 'open'), ('foo', 'ohlc', 'high'),
-        ('foo', 'ohlc', 'low'), ('foo', 'ohlc', 'close'),
-        ('foo', 'sum', 'foo'))), index=pd.date_range(
+        ('foo', 'sum', 'foo'), ('foo', 'ohlc', 'open'),
```
So the only change in behavior is that the output order is now (consistently) the arguments passed to `.agg`?

Since we did `['sum', 'ohlc']` the output order is `'sum', 'open', 'high', 'low', 'close'`?

If so, let's add an entry to the release notes, under the Groupby bug fix section, saying that `.agg` now respects the order of arguments when `'ohlc'` is one of the aggfuncs.

WillAyd replied:
This might be rather difficult to express in a whatsnew since it's really just applicable to `_aggregate_multiple_funcs` here. 

Perhaps a follow up to consistently use OrderedDict's in the groupby module is in order?

TomAugspurger replied:
Is `_aggregate_multiple_funcs` called by just by functions like `ohlc` that have multiple outputs per input? Or is it also called by `.agg([funcs])`?

WillAyd replied:
The latter. So this has implications to ordering which previously would have been non-deterministic in Py35. 

So I think it might make sense as a follow up to use a OrderedDict and make better guarantees about the returned column order, but don't think that should hold this one up

WillAyd replied:
See #25692 as follow up

TomAugspurger reviewed (commented):
LGTm, aside from a small request in the whatsnew.

TomAugspurger commented:
Sounds good.

On Tue, Mar 12, 2019 at 12:46 PM William Ayd <notifications@github.com>
wrote:

> *@WillAyd* commented on this pull request.
> ------------------------------
>
> In pandas/tests/groupby/test_groupby.py
> <https://github.com/pandas-dev/pandas/pull/25224#discussion_r264805394>:
>
> > @@ -1690,14 +1690,11 @@ def test_groupby_agg_ohlc_non_first():
>          [1, 1, 1, 1, 1],
>          [1, 1, 1, 1, 1]
>      ], columns=pd.MultiIndex.from_tuples((
> -        ('foo', 'ohlc', 'open'), ('foo', 'ohlc', 'high'),
> -        ('foo', 'ohlc', 'low'), ('foo', 'ohlc', 'close'),
> -        ('foo', 'sum', 'foo'))), index=pd.date_range(
> +        ('foo', 'sum', 'foo'), ('foo', 'ohlc', 'open'),
>
> The latter. So this has implications to ordering which previously would
> have been non-deterministic in Py35.
>
> So I think it might make sense as a follow up to use a OrderedDict and
> make better guarantees about the returned column order, but don't think
> that should hold this one up
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/25224#discussion_r264805394>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIg6jzfNZb8_yQW9MXzmjZPYWaO0Cks5vV-gIgaJpZM4attTh>
> .
>

WillAyd reviewed (approved):
lgtm

alexander-ponomaroff commented:
Thank you @WillAyd and @TomAugspurger for the feedback and the help on this one. @jreback could you take a look if there are any additional changes that need to take place here?

jreback commented:
merging if any additional whatsnew is needed, can do in #25962
